### PR TITLE
underline nested type variable upon type-class constraint violation

### PR DESCRIPTION
**Pull request by davidchambers**
Before:

    concatMaybes :: Semigroup a => Maybe a -> Maybe a -> Maybe a
                    ^^^^^^^^^^^    ^
                                   1

After:

    concatMaybes :: Semigroup a => Maybe a -> Maybe a -> Maybe a
                    ^^^^^^^^^^^          ^
                                         1

Does this change look okay to you, @kedashoe?

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -830,13 +838,14 @@
                       reprs.join(_(', ')));
 
     var padding = _(showTypeSig_(expTypes.slice(0, index)));
```
Comment (kedashoe) [inline index.js]: I believe this line should be something like

`var padding = _(showTypeSig_(expTypes.slice(0, isNaN(index) ? expTypes.length - 1 : index)));`

to properly point to type variables in the return value.

Comment (kedashoe) [inline index.js]: (also might want to add a test for pointing to return values, which is how I discovered something might be amiss..)

File: index.js [hunk 2]
```diff
@@ -830,13 +838,14 @@
...
+    var f = underline(typePath[0])(propPath);
 
     return new TypeError(unlines([
```
Comment (kedashoe) [inline index.js]: I know I already said it, but I love the formatting of these now with the longer lines + `unlines` :)

File: index.js (hunk 1 above)
Comment (davidchambers) [inline index.js]: Great catch! I'll add a failing test case then make your suggested fix.

File: index.js (hunk 2 above)
Comment (davidchambers) [inline index.js]: :smile:

File: index.js (hunk 1 above)
Comment (davidchambers) [inline index.js]: :zap:

File: index.js [hunk 3]
```diff
@@ -1053,9 +1066,10 @@
                       constraints,
                       expArgTypes.concat([expRetType]),
                       values[idx],
-                      index,
+                      isNaN(index) ? expArgTypes.length : index,
```
Comment (davidchambers) [inline index.js]: How does this look, Kevin? It seems preferable to handle `NaN` sooner rather than later.

Comment (davidchambers) [inline index.js]: Actually, passing around `expTypes` rather than `expArgTypes` and `expRetType` turns out to be simpler. `NaN` is no longer necessary. :)

Comment (kedashoe) [inline index.js]: Nice change, looks good!

File: index.js [hunk 4]
```diff
@@ -831,12 +839,17 @@
 
     var padding = _(showTypeSig_(expTypes.slice(0, index)));
 
+    //  f :: (String -> String) -> String
+    var f = function(f) {
+      return trimTrailingSpaces(underline(typePath[0], propPath, f));
+    };
+
```
Comment (davidchambers) [inline index.js]: While updating Sanctuary I noticed that I'd forgotten to apply `trimTrailingSpaces` below. I updated the definition of `f` accordingly. There's now no reason to curry `underline`, so I reverted that change.

File: index.js [hunk 5]
```diff
@@ -158,7 +158,7 @@
 
   //  trimTrailingSpaces :: String -> String
   var trimTrailingSpaces = function(s) {
-    return s.replace(/[ ]+$/, '');
+    return s.replace(/[ ]+$/gm, '');
   };
```
Comment (davidchambers) [inline index.js]: Trimming all trailing spaces *after* formatting a message simplifies matters as we no longer need to clean up as we go.

File: index.js [hunk 6]
```diff
@@ -1163,18 +1195,16 @@
...
                 throw invalidValue(name,
                                    constraints,
-                                   expArgTypes.concat([expRetType]),
+                                   expTypes,
                                    value,
                                    index,
                                    determineActualTypesLoose([value]));
               }
+              satisfactoryTypes(name, constraints, expTypes,
+                                $typeVarMap, value, index);
```
Comment (davidchambers) [inline index.js]: It turns out this function is only being called for side effects! At least now it's clear that this is the case.

File: index.js (hunk 5 above)
Comment (davidchambers) [inline index.js]: #46

**Pull request merged**